### PR TITLE
Levelling v2: resume/migration CPU checks

### PR DESCRIPTION
**Pull request by robhoes**

### CODE REVIEW

**Comment by robhoes** [comment]
Travis tests failed, but due to something unrelated it seems.

File: ocaml/xapi/cpuid_helpers.ml [hunk 1]
```diff
@@ -143,31 +114,55 @@ let update_cpu_flags ~__context ~vm ~host =
...
+		let vm_cpu_vendor = List.assoc cpu_info_vendor_key vm_cpu_info in
+		debug "VM last booted on CPU of vendor %s; destination's CPUs are of vendor %s" vm_cpu_vendor dest_cpu_vendor;
+		if vm_cpu_vendor <> dest_cpu_vendor then
+			fail "VM last booted on a host which had a CPU from a different vendor."
+	end;
+	if List.mem_assoc cpu_info_features_key vm_cpu_info then begin
+		(* Check the VM was last booted on a CPU whose features are a subset of the features of this destination's CPU. *)
+		let vm_cpu_features = List.assoc cpu_info_features_key vm_cpu_info in
+		debug "VM last booted on CPU with features %s; destination's CPUs have features %s" vm_cpu_features dest_cpu_features;
+		let vm_cpu_features' = vm_cpu_features |> features_of_string in
```
Comment (robhoes) [inline ocaml/xapi/cpuid_helpers.ml]: Oops, forgot to apply `upgrade_features` here.

File: ocaml/xapi/xapi_vm_migrate.ml [hunk 2]
```diff
@@ -788,7 +788,7 @@ let assert_can_migrate  ~__context ~vm ~dest ~live ~vdi_map ~vif_map ~options =
 	in
 	match migration_type with
 	| `intra_pool host ->
-		if (not force) && live then Cpuid_helpers.assert_vm_is_compatible ~__context ~vm ~host ();
+		if (not force) && live then Cpuid_helpers.assert_vm_is_compatible_with_host ~__context ~vm ~host ();
```
Comment (robhoes) [inline ocaml/xapi/xapi_vm_migrate.ml]: This does not yet take into account the SXM-across-versions case.

File: ocaml/xapi/xapi_vm_migrate.ml [hunk 3]
```diff
@@ -800,7 +800,7 @@ let assert_can_migrate  ~__context ~vm ~dest ~live ~vdi_map ~vif_map ~options =
 				"VIF mapping is not supported for intra-pool migration"]))
 	| `cross_pool remote_rpc ->
 		if (not force) && live then
-			Cpuid_helpers.assert_vm_is_compatible ~__context ~vm ~host:dest_host_ref
+			Cpuid_helpers.assert_vm_is_compatible_with_pool ~__context ~vm
```
Comment (robhoes) [inline ocaml/xapi/xapi_vm_migrate.ml]: Neither does this.

File: scripts/xapi.conf [hunk 4]
```diff
@@ -68,6 +68,9 @@ igd-passthru-vendor-whitelist = 8086
 # kinds of Intel hardware.
 # gvt-g-whitelist = /etc/gvt-g-whitelist
 
+# Mask of CPU features that should not prevent a VM migration
```
Comment (euanh) [inline scripts/xapi.conf]: Can you comment on why base_ecx.7 is ignored?

Comment (robhoes) [inline scripts/xapi.conf]: It's for historical reasons really. I have filled in here what was the `cpuid_default_feature_mask`, which was set for compatibility with XS versions older than 5.6, which had different pool join checks. In any case, this mask needs to be replaced with the real don't care mask, once we know what it is.

**Comment by robhoes** [comment]
@euanh I've done a rebase after merging your PR.

**Comment by euanh** [comment]
:shipit:

**Pull request merged**